### ISSUE TITLE
Defaulting `CLANG_ENABLE_OPAQUE_POINTERS_INTERNAL` to `ON` broke OpenCL

### ISSUE THREAD

**Issue by illwieckz** [llvm/llvm-project#54970]
Mesa Clover relying on Clang and libclc started to fail some days ago when building against LLVM `main` branch. I bisected the LLVM tree and I noticed the error appeared with LLVM commit 702d5de4380b1e1554e5b90863093c3a57f76f70 which sets `CLANG_ENABLE_OPAQUE_POINTERS_INTERNAL` to `ON`:

> [Clang] Enable opaque pointers by default
>
> Enable opaque pointers by default in clang, which can be disabled
> either via cc1 option -no-opaque-pointers or cmake flag
> -DCLANG_ENABLE_OPAQUE_POINTERS=OFF.
>
> See https://llvm.org/docs/OpaquePointers.html for context.
>
> Differential Revision: https://reviews.llvm.org/D123300

Starting with this commit, when running [Luxmark 3.1](http://wiki.luxcorerender.org/LuxMark_v3#Binaries) OpenCL benchmark, I get this crash:

```
luxmark.bin: …/include/llvm/IR/Type.h:381: llvm::Type *llvm::Type::getNonOpaquePointerElementType() const: Assertion `NumContainedTys && "Attempting to get element type of opaque pointer"' failed.
```
I build LLVM and Mesa using my [`user-mesa`](https://gitlab.com/illwieckz/i-love-compute/-/blob/master/scripts/user-mesa) build script that rebuilds Clang, libclc, Mesa and other components, so every time I rebuilt LLVM I also rebuilt Mesa (not from scratch, Meson rebuilding what it detects to have to be rebuilt).

I don't know if the bug should be addressed in LLVM or in Mesa.

**Comment by tstellar** [comment]
@illwieckz This is likely a mesa bug, I would recommend filing a bug against that project.

**Comment by illwieckz** [comment]
@tsellar: done: https://gitlab.freedesktop.org/mesa/mesa/-/issues/6342

**Comment by illwieckz** [comment]
Note that building Mesa against current LLVM `main` branch is also broken by another LLVM commit, commit 93471e65df48372ee59bd0c2f8ba58a254ba1ca5, see [`mesa/mesa#6333`](https://gitlab.freedesktop.org/mesa/mesa/-/issues/6333).

To debug this issue on Mesa side, one would want to build against LLVM commit 63f2d1f4d4b8ee284b4ab977242e322a9458a168 or use the code in my merge request [`mesa/mesa!16002`](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/16002).

**Comment by nikic** [comment]
Yes, this needs to be addressed on the mesa side, and `-DCLANG_ENABLE_OPAQUE_POINTERS=OFF` can be used in the meantime to keep the build working.

From a quick grep, https://gitlab.freedesktop.org/mesa/mesa/-/blob/ff05137c2dbf9c6b8e5d4ce31063516e5c04569f/src/gallium/frontends/clover/llvm/codegen/common.cpp#L207 needs to use `getParamByValType()`, while https://gitlab.freedesktop.org/mesa/mesa/-/blob/main/src/gallium/frontends/clover/llvm/codegen/common.cpp#L218 needs to use `getParamAlign()` -- possibly combined with some frontend change to place the align attribute (clang places it for OpenCL kernels, but I don't know how mesa and clang interact here).

**Comment by illwieckz** [comment]
@nikic thanks for the answer. I noticed I can make Clover passes `-no-opaque-pointer` option to Clang to workaround the issue, while still building default LLVM with `CLANG_ENABLE_OPAQUE_POINTERS` enabled.

See: https://gitlab.freedesktop.org/mesa/mesa/-/issues/6342#note_1344480

For fixing the root cause of this I would probably have to wait for a Mesa wizard to look at it.
